File: backend/src/main/java/com/gxs/bank/controller/BillPaymentController.java
package com.gxs.bank.controller;

import com.gxs.bank.dto.response.ApiResponse;
import com.gxs.bank.model.BillPayment;
import com.gxs.bank.model.User;
import com.gxs.bank.service.BillPaymentService;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;

import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.UUID;

@RestController
@RequestMapping("/api/bills")
@RequiredArgsConstructor
public class BillPaymentController {
    private final BillPaymentService billPaymentService;

    @PostMapping("/pay")
    public ResponseEntity<ApiResponse<BillPayment>> payBill(@AuthenticationPrincipal User user, @RequestBody Map<String, Object> payload) {
        BillPayment request = new BillPayment();
        request.setBillerName((String) payload.get("billerName"));
        request.setBillerCategory(BillPayment.BillerCategory.valueOf((String) payload.get("billerCategory")));
        String billerAccNo = (String) payload.get("billerAccountNumber");
        if (billerAccNo == null) billerAccNo = (String) payload.get("consumerNumber");
        request.setBillerAccountNumber(billerAccNo != null ? billerAccNo : "N/A");
        request.setAmount(new java.math.BigDecimal(payload.get("amount").toString()));
        UUID accountId = UUID.fromString((String) payload.get("accountId"));
        
        BillPayment payment = billPaymentService.payBill(user.getId(), accountId, request);
        return ResponseEntity.ok(ApiResponse.ok("Bill payment successful", payment));
    }

    @GetMapping("/history")
    public ResponseEntity<ApiResponse<List<BillPayment>>> getHistory(@AuthenticationPrincipal User user) {
        return ResponseEntity.ok(ApiResponse.ok(billPaymentService.getHistory(user.getId())));
    }

    @GetMapping("/billers")
    public ResponseEntity<ApiResponse<List<String>>> getBillers() {
        List<String> categories = Arrays.stream(BillPayment.BillerCategory.values()).map(Enum::name).toList();
        return ResponseEntity.ok(ApiResponse.ok(categories));
    }
}

File: backend-python/com/gxs/bank/controller/BillPaymentController.py
from decimal import Decimal

from fastapi import APIRouter, Depends

from com.gxs.bank.config.database import get_db
from com.gxs.bank.config.serialization import serialize
from com.gxs.bank.dto.response.ApiResponse import ok
from com.gxs.bank.model.BillPayment import BillPayment
from com.gxs.bank.security.JwtAuthenticationFilter import get_current_user
from com.gxs.bank.service.BillPaymentService import BillPaymentService


router = APIRouter(prefix="/api/bills", tags=["BillPayment"])


@router.post("/pay")
def payBill(payload: dict, current_user=Depends(get_current_user), db=Depends(get_db)):
    request = BillPayment()
    request.billerName = payload.get("billerName")
    request.billerCategory = BillPayment.BillerCategory(payload.get("billerCategory"))
    billerAccNo = payload.get("billerAccountNumber")
    if billerAccNo is None:
        billerAccNo = payload.get("consumerNumber")
    request.billerAccountNumber = billerAccNo if billerAccNo is not None else "N/A"
    request.amount = Decimal(str(payload.get("amount")))
    accountId = payload.get("accountId")

    payment = BillPaymentService(db).payBill(current_user.id, accountId, request)
    return ok(serialize(payment), "Bill payment successful")


@router.get("/history")
def getHistory(current_user=Depends(get_current_user), db=Depends(get_db)):
    history = BillPaymentService(db).getHistory(current_user.id)
    return ok(serialize(history))


@router.get("/billers")
def getBillers():
    categories = [value.value for value in BillPayment.BillerCategory]
    return ok(categories)
